### PR TITLE
[5.5] Eloquent object casting

### PR DESCRIPTION
**Pull request by tillkruss**
Re-implementation of #13706 after a long discussion with @adamwathan and @JosephSilber.

```php
$user = User::create([
    'name' => 'Taylor',
    'token' => new Secret('Adam123'),
]);

$user->token; // Secret instance
$user->token->secret; // "Adam123"

$user->price = (new Money(10))->setCurrency('BTC');

$user->price; // Money instance
$user->price->amount; // "10"
$user->toArray()['price']; // "10"
$user->currency; // "BTC"
```

Example classes are in the next comment below.

### CODE REVIEW

**Comment by tillkruss** [comment]
```php
class User extends Model
{
    protected $casts = [
        'token' => 'secret',
        'price' => 'money:price,currency',
    ];

    public function castToSecret($value)
    {
        return Secret::fromHash($value);
    }

    // no "castFromSecret" here, because Secret has a __toString() method

    public function castToMoney($amount, $currency)
    {
        return (new Money($amount))->setCurrency($currency);
    }

    public function castFromMoney($money)
    {
        return [$money->amount, $money->currency];
    }
}
```

```php
class Secret
{
    public $secret;

    public function __construct($secret)
    {
        $this->secret = $secret;
    }

    public static function fromHash($secret)
    {
        return new static(base64_decode($secret));
    }

    public function __toString()
    {
        return base64_encode($this->secret);
    }
}
```

```php
class Money
{
    public $amount;

    public $currency;

    public function __construct($amount)
    {
        $this->amount = $amount;
    }

    public function setCurrency($currency)
    {
        $this->currency = $currency;

        return $this;
    }
}
```

**Comment by devcircus** [comment]
glad to see this come back to life

**Comment by joshmanders** [comment]
This is going to be awesome!

**Comment by sisve** [comment]
Relevant, and ignored, discussion: https://github.com/laravel/internals/issues/9

**Comment by sisve** [comment]
1. Why isn't this implemented as separate "casting objects". An example; a CastsMoney that has castToPHP and castToDatabase? This would also give the possibility to expose data on how many columns are expected.
2. How are we expected to handle casting of null values? castAttribute is doing a null-check before even calling castToClass.
3. How is the casting logic expected to expose any dependencies needed? Constructor injection is common in other places in the framework, this can't be done here. Method injections, like controller actions?

**Comment by sisve** [comment]
Couldn't the previous code in castAttribute use the new logic to? castToInt, castToFloat, castToString, castToBool, ...? Or perhaps just the non-scalar values like castToCollection, castToJson, castToDate, castToDateTime, castToTimestamp?

**Comment by RemiCollin** [comment]
Does the object properties have to be public ?

We'd moslty want them `protected` for most value objects.

And +1 with sisve about dependency injection. It should be handled.

**Comment by tillkruss** [comment]
@RemiCollin: No, the properties' visibility doesn't matter, because your `castFrom*()` method can get the model attributes however it pleases.

**Comment by tillkruss** [comment]
@sisve

> How are we expected to handle casting of null values? castAttribute is doing a null-check before even calling castToClass.

`$user->price = null;` will unset the `price` and `currency` attribute on the model.

If you want to handle `null` scenarios differently, you can set `$user->price = new Money(null)` and let the object decide what to do/return.

> How is the casting logic expected to expose any dependencies needed? Constructor injection is common in other places in the framework, this can't be done here. Method injections, like controller actions?

You can use Laravel's `resolve()` helper in your `castTo*()` methods.

**Comment by sisve** [comment]
When I'm talking about null values I am thinking of null values in the database. How can I transform them into a meaningful value in php? The castAttribute method checks for nulls, and return nulls, before calling the new casting code. I would prefer if my castToShazaam() method would receive the null value and handle it itself.

I don't like the pattern of calling resolve() or App::make() from within models (or any other code), it hides dependencies. I prefer it if method has a clear scope and intent, and only uses what has been provided; not call out and introduce dependencies to things that aren't obvious.

**Comment by RemiCollin** [comment]
@tillkruss : ok I see, at first I thought the intention was to handle this in base `Model` class for auto magic. But being verbose here makes sense, for flexibility.

For DI, agree with @sisve that calling `resolve()` would obviously be a code smell. Why not having the container as a property of the `Model` class, so we can make calls like $this->app->make(Money::class). ?

**Comment by tillkruss** [comment]
@sisve: All `null` database values are passed to the `castTo*()` method, so you can decide what to do with them.

If that's not the case for you when testing this PR locally, please provide a working code sample for me to test against.

**Comment by sisve** [comment]
I'm looking at the github source view, and it clearly shows the is_null check before doing calling castToClass. https://github.com/tillkruss/framework/blob/36c982ea23bd6759acfd088ad9972bcf16f813f7/src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php#L474-L480

What you describe matches https://github.com/laravel/framework/commit/b36602b9796ab42f914c91d608f062cd8d2f8246 but that isn't part of this PR, according to the GitHub interface.

**Comment by sisve** [comment]
Can we change mergeAttributesFromClassCasts into preferring the castFrom* methods instead of __toString? I think that if I write a castFromShazaam method, it should have priority over my Shazaam::__toString() may have another focus than giving values to persist into the database.

**Comment by tillkruss** [comment]
@sisve: That makes sense, just switched it around. Thanks!

Regarding the null-scenario: When the model is instantiated from a database record, the `castAttribute()` method is **not** called and `castTo*()` will receive whatever database value(s).

**Comment by JosephSilber** [comment]
@sisve 
> How is the casting logic expected to expose any dependencies needed? Constructor injection is common in other places in the framework, this can't be done here.

Value objects generally don't (or shouldn't) have dependencies. If you do need them, use the `resolve` method from within your `castTo*` method.

> I don't like the pattern of calling resolve() [...], it hides dependencies. I prefer it if method has a clear scope and intent, and only uses what has been provided; not call out and introduce dependencies to things that aren't obvious.

These `castTo*` are factory methods. If they need external dependencies, they have to pull it in somehow. There's no going around that, and using `resolve` is a very elegant approach.

@RemiCollin 

> For DI [...] calling resolve() would obviously be a code smell.

Uh... really? Why? Calling `resolve` from within a factory method means the code is still fully, 100% testable. Don't let people scare you away from practical solutions.

> Why not having the container as a property of the Model class, so we can make calls like `$this->app->make(Money::class)`?

And how would that be any better? You're reaching into the very same container, and resolving a class dynamically.

---

Dynamic DI is a very useful feature. Use it when applicable and stay happy :grinning:

**Comment by RemiCollin** [comment]
@JosephSilber : you're right the result is the same, I didn't see it from that angle. About Value Objects not having dependencies, I'd agree 99%, though sometime it's nice to have this possibility. 

Also, while reading the examples above, I wonder if it wouldn't be better to explicitely return `key` `value` pairs from the `castFrom*` method. I feel it could become error prone to only rely on the attribute's order if the number of attributes is important. 

Something like this maybe: 

```php 

public function castFromMoney($money)
{
    return [
        'price' => $money->amount, 
        'currency' => $money->currency,
    ];
 }

```

**Comment by JosephSilber** [comment]
@RemiCollin those column names shouldn't be hard coded. The casting method should not be aware of what columns the value is stored in.

Imagine you have 2 money values in the same table:

- price
- price_currency
- shipping
- shipping_currency

Then you'd have this in your casts:

```php
$casts = [
    'price' => 'money:price,price_currency',
    'shipping' => 'money:shipping,shipping_currency',
];
```

As you can see, the actual column names have to stay abstract. It's up to the user to specify the columns.

File: src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php [hunk 1]
```diff
@@ -485,8 +505,81 @@ protected function castAttribute($key, $value)
...
+
+                $attributeCount = count($attributes);
+                $object = $this->castToClass($attribute);
+
+                $castedAttributes = method_exists($this, 'castFrom'.Str::studly($type))
+                    ? $this->{'castFrom'.Str::studly($type)}($object)
+                    : $object->__toString();
+
+                if ($attributeCount !== count($castedAttributes)) {
+                    throw new LogicException("Class cast {$attribute} must return {$attributeCount} attributes");
```
Comment (jmarcher) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Wouldn't be better to change this to:
"Class cast {$attribute} must return {$attributeCount} attribute(s)" ? I've seen it in some parts of the core

Comment (tillkruss) [inline src/Illuminate/Database/Eloquent/Concerns/HasAttributes.php]: Yep, just did that. Thanks!

**Comment by ConnorVG** [comment]
This is marvellous; I'd be very eager to read the documentation.

**Comment by jimgwhit** [comment]
Just curious, why the over thinking of dealing with money? I have always just stored a 0.00 if there is no value.

**Comment by tillkruss** [comment]
It's just an example.

**Comment by ConnorVG** [comment]
What level of SQL query-ability would we have with this? Based on the money example, I would love to do something like:
`User::where('money.currency', 'USD')` etc

**Comment by jmarcher** [comment]
I don't believe that could be possible.
This pr is about casting not extending Laravel relationships

**Comment by ConnorVG** [comment]
@jmarcher I know what the target of the PR is, the problem is that the data is **very limited** if we have no way of comparing it during querying.

**Comment by sisve** [comment]
@ConnorVG it would be hard to implement what you're asking about, querying a part of the value object. Something would have to know how to translate `money.currency` to the proper database column, such functionality does not exist in this PR.

However, it would seem _more possible_ to implement `User::where('money', new Money(1337, 'USD'))` since the model has the ability to convert the Money object into the necessary database columns. Such statement would then generate `WHERE (price = 1337 AND price_currency = 'USD')`. This sounds like something that would be possible with the current setup.

But this opens up more questions, how about comparisons? Should you be able to filter rows with than 1337 USD? How would that handle other currencies? This would easily expand the scope of this issue enormously.

**Comment by ConnorVG** [comment]
Yeah, I understand the issue I just noticed no one seems to have pointed it out. A simple *STRICT* check shouldn't be too hard if we enforce (or at-least highly recommend) JSON column types then make it driver based (based on the active connection). It's non-strict that would be a complete PITA.

I'd be less interested in checking the full object, hence my example only querying for the currency.

**Comment by tillkruss** [comment]
If you choose to use a JSON column to store your data, you can just do `User::where('money->currency', 'USD')`.

**Comment by DanielCoulbourne** [comment]
I feel like I'm missing something. Why couldn't this just be an accessor?

**Comment by linaspasv** [comment]
What's the status of the PR? When can we expect this could be merged to the core?

**Comment by tillkruss** [comment]
@DanielCoulbourne: Duplication for example, what if you model has 3 attributes that are all the same object, for example: `price`, `sale_price`, `wholesale_price`. Instead of having 6 methods, you only need two.

**Comment by garygreen** [comment]
@tillkruss I too am struggling to see why all of this can't just be a simple accessor. For example:

```php
<?php

class User {

    public function getSecretAttribute()
    {
        return $this->buildOrGetInstance('secret', function() {
            return new Secret($this->attributes['secret']);
        });
    }

    public function setSecretAttribute($secretValue)
    {
        $this->secret->setSecret($secretValue);
        $this->attributes['secret'] = $secretValue;
    }

}
```

The `buildOrGetInstance` is just a simple function on your base model that will get the instance by name (similar to `getRelation($name)`) or call the closure and create it, caching it under `->instances`

This will straight away with with `toArray` and everything.

My thoughts are just kinda "meh" towards this when we've already got something similar. Seems pretty over-engineered.

**Comment by linaspasv** [comment]
@garygreen I like your pseudocode. Is there any available implementation of buildOrGetInstance? I guess the feature/possibility of accessor caching is missing in current eloquent object implementation.

On the other hand @tillkruss implementation focuses more on a possibility of creating custom casts. As of custom casts - maybe it should be in a similar style like we currently have for custom validation rules creation ``Validator::extend`` . It would make sense to have custom casts defined globally instead of re-defining on each model separately.

**Comment by garygreen** [comment]
>  I guess the feature/possibility of accessor caching is missing in current eloquent object implementation.

That's true, to be honest I was kinda suprised to find out recently that when you access attributes that are cast to `Carbon` new instances of `Carbon` are created every time you access an attribute.

E.g.:

```php
$user->created_at; // Creates new carbon model....
$user->created_at; // Creates 2nd carbon model...
$user->created_at; // Creates 3rd carbon model... etc
```

Seems a bit of a waste to keep creating new instances, but I guess it's useful if you want to mutate the date in some way... then again you can always use `->created_at()->copy()` to guarantee you've got yourself a new mutable instance.

**Comment by unstoppablecarl** [comment]
I downloaded the branch and tested it out. I have some serious problems with this so far. I'm not sure if I'm getting what this solves, maybe the example provided is not a good use case. 

## Problems:

### 1) Creates Variant interface for attributes

```php
$user = new User();
$user->price->amount; // property of non object error
$user->price->currency; // property of non object error
```

`$user->price` is not an instance of `Money` until you set it, so the interface for retrieving the `price` attribute is now variant.

```php
if($user->price instanceof Money){
	$user->price->amount;
} else {
	$user->price ?: 0;
}
```

### 2) Incomplete Symmetry

There are now 2 seemingly valid places to set the currency (which is fine) but they are not synced both ways.

```php
$user->price = (new Money(10))->setCurrency('BTC');
$user->currency = 'USD';
$user->price->currency; // BTC, does not sync from attribute to value object
```

```php
$user->price = (new Money(10))->setCurrency('BTC');
$user->price->currency = 'USD';
$user->price->currency; // USD
$user->currency; // USD, seems to sync from value object to attribute

// non-public internally referenced `$attributes` property is not updated
$user->attributes['currency'] // 'BTC'
```

Anything internally referencing the attributes array is now getting incorrect data.

**Comment by garygreen** [comment]
@unstoppablecarl I completely agree with both points. Did you see my example above? It solves both those problems with simple getters and setters if you implement a very basic `buildOrGetInstance`

**Comment by taylorotwell** [comment]
I think this is just more than I want to take on right now.

**Comment by a-komarev** [comment]
It would be great to have this functionality out of the box in Laravel 5.7

**Comment by regularlabs** [comment]
I ended up using this on my models via a trait:
```
    protected function castAttribute($key, $value)
    {
        if (isset($this->casts[$key]) && $this->casts[$key] == 'datetime') {
            return MyOwnCarbon($value);
        }

        return parent::castAttribute($key, $value);
    }
```

Then in my models I have $casts, like:
```
    protected $casts = [
        'date_time_start'        => 'datetime',
        'date_time_finish'       => 'datetime',
    ];
```

**Comment by antonkomarev** [comment]
It could be extracted to trait then.

**Pull request closed without merging**